src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorNodeWriter.cs

```diff
@@ -49,5 +53,169 @@ public sealed override void WriteCSharpCodeAttributeValue(CodeRenderingContext c
             context.Diagnostics.Add(BlazorDiagnosticFactory.Create_CodeBlockInAttribute(node.Source, content));
             return;
         }
+
+
+        // Currently the same for design time and runtime
+        public void WriteComponentTypeInferenceMethod(CodeRenderingContext context, ComponentTypeInferenceMethodIntermediateNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // This is ugly because CodeWriter doesn't allow us to erase, but we need to comma-delimit. So we have to
+            // materizalize something can iterate, or use string.Join. We'll need this multiple times, so materializing
+            // it.
+            var parameters = GetParameterDeclarations();
+
+            // This is really similar to the code in WriteComponentAttribute and WriteComponentChildContent - except simpler because
+            // attributes and child contents look like variables. 
+            //
+            // Looks like:
+            //
+            //  public static void CreateFoo_0<T1, T2>(RenderTreeBuilder builder, int seq, int __seq0, T1 __arg0, int __seq1, global::System.Collections.Generic.List<T2> __arg1, int __seq2, string __arg2)
+            //  {
+            //      builder.OpenComponent<Foo<T1, T2>>();
+            //      builder.AddAttribute(__seq0, "Attr0", __arg0);
+            //      builder.AddAttribute(__seq1, "Attr1", __arg1);
+            //      builder.AddAttribute(__seq2, "Attr2", __arg2);
+            //      builder.CloseComponent();
+            //  }
+            //
+            // As a special case, we need to generate a thunk for captures in this block instead of using
+            // them verbatim.
+            //
+            // The problem is that RenderTreeBuilder wants an Action<object>. The caller can't write the type
+            // name if it contains generics, and we can't write the variable they want to assign to. 
+            var writer = context.CodeWriter;
+
+            writer.Write("public static void ");
```

Comment:
Is this a method-within-a-method? If so, really handy that the compiler now allows this!

Reply:
Well you can use local functions, but this is not. Local functions would be fraught with peril for our use case.

This is a separate namespace/class/method from your original code. The main reason why its a separate type is so that we can avoid renaming the generic type parameters. If the thunk is part of the same class, then it shares the type parameters of the calling component, which is very counterproductive.